### PR TITLE
Add View License Returns page

### PR DESCRIPTION
**Pull request by jonathangoulding**
The existing service handling view license is slow because it loads all the data for the tabs in one render. Work has been done previously to refactor the summary page to load only the summary information.

This change will introduce a returns controller, service and presenter to handle the view license returns page.

This will share he same view as the summary page and load the same 'common data' established in previous work.

Both pieces of work are part of - https://eaflood.atlassian.net/browse/WATER-4444

Previous work as mentioned above - https://github.com/DEFRA/water-abstraction-system/pull/957

### CODE REVIEW

**Comment by jonathangoulding** [comment]
Need to implement the no returns tab / view.

File: app/routes/licence.routes.js [hunk 1]
```diff
@@ -8,9 +8,18 @@ const routes = [
     path: '/licences/{id}/summary',
     handler: LicencesController.viewSummary,
     options: {
-      description: 'View a licence page'
+      description: 'View a summary licence page'
```
Comment (robertparkinson) [inline app/routes/licence.routes.js:11]: shouldn't this be `View a licence summary page` ?

File: app/routes/licence.routes.js [hunk 2]
```diff
@@ -8,9 +8,18 @@ const routes = [
...
     }
-  }, {
+  },
+  {
+    method: 'GET',
+    path: '/licences/{id}/returns',
+    handler: LicencesController.viewReturns,
+    options: {
+      description: 'View a returns licence page'
```
Comment (robertparkinson) [inline app/routes/licence.routes.js:19]: same here, `View a licence returns page`

File: app/controllers/licences.controller.js [hunk 3]
```diff
@@ -6,8 +6,8 @@
  */
 
 const InitiateReturnRequirementSessionService = require('../services/return-requirements/initiate-return-requirement-session.service.js')
-const ViewLicenceSummaryService = require('../services/licences/view-license-summary.service')
-
+const ViewLicenceSummaryService = require('../services/licences/view-licence-summary.service')
+const ViewLicenceReturnsService = require('../services/licences/view-licence-returns.service')
```
Comment (Cruikshanks) [inline app/controllers/licences.controller.js:10]: NIT - alphabetical ordering and adding some white space

```suggestion
const ViewLicenceReturnsService = require('../services/licences/view-licence-returns.service')
const ViewLicenceSummaryService = require('../services/licences/view-licence-summary.service')

```

File: app/controllers/licences.controller.js [hunk 4]
```diff
@@ -35,8 +35,20 @@ async function viewSummary (request, h) {
...
+    ...data
+  })
+}
+
 module.exports = {
   noReturnsRequired,
   returnsRequired,
-  viewSummary
+  viewSummary,
+  viewReturns
```
Comment (Cruikshanks) [inline app/controllers/licences.controller.js:53]:
```suggestion
  viewReturns,
  viewSummary
```

File: app/services/licences/view-licence-returns.service.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * Orchestrates fetching and presenting the data needed for the licence summary page
+ * @module ViewLicenceSummaryService
+ */
+
+const FetchLicenceReturnsService = require('./fetch-licence-returns.service')
+const ViewLicenceService = require('./view-licence.service')
+const ViewLicenceReturnsPresenter = require('../../presenters/licences/view-licence-returns.presenter')
+const PaginatorPresenter = require('../../presenters/paginator.presenter')
+/**
```
Comment (robertparkinson) [inline app/services/licences/view-licence-returns.service.js:12]: alan is going to point out a NIT here of an extra empty line between the declarations and the comments

File: app/presenters/licences/view-licence-returns.presenter.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Formats data for common licence data `/licences/{id}` page's
+ * @module ViewLicenceReturnsPresenter
+ */
+
+const { formatLongDate } = require('../base.presenter')
+
+/**
+ * Formats data for common licence data `/licences/{id}` page's
```
Comment (Cruikshanks) [inline app/presenters/licences/view-licence-returns.presenter.js:11]: NIT

```suggestion
 * Formats common data for the `/licences/{id}/*` view licence pages
```

File: app/presenters/licences/view-licence-returns.presenter.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
+'use strict'
+
+/**
+ * Formats data for common licence data `/licences/{id}` page's
```
Comment (Cruikshanks) [inline app/presenters/licences/view-licence-returns.presenter.js:4]: NIT

```suggestion
 * Formats common data for the `/licences/{id}/*` view licence pages
```

File: app/presenters/licences/view-licence-returns.presenter.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+      id: r.id,
+      reference: r.returnReference,
+      purpose: _formatPurpose(r.metadata.purposes),
+      description: r.metadata.description,
+      dueDate: formatLongDate(new Date(r.dueDate)),
+      status: _formatStatus(r.status),
+      dates: `${formatLongDate(new Date(r.startDate))} to ${formatLongDate(new Date(r.endDate))}`
+    }
+  })
+}
```
Comment (Cruikshanks) [inline app/presenters/licences/view-licence-returns.presenter.js:48]: NIT - ordering and some whitespace

```suggestion
function _formatReturnToTableRow (returns) {
  return returns.map(r => {
    return {
      id: r.id,
      reference: r.returnReference,
      purpose: _formatPurpose(r.metadata.purposes),
      description: r.metadata.description,
      dueDate: formatLongDate(new Date(r.dueDate)),
      status: _formatStatus(r.status),
      dates: `${formatLongDate(new Date(r.startDate))} to ${formatLongDate(new Date(r.endDate))}`
    }
  })
}

function _formatStatus (status) {
  if (status === 'completed') return 'COMPLETE'
  if (status === 'due') return 'OVERDUE'
  
  return 'NO STATUS'
}
```

File: app/presenters/licences/view-licence-returns.presenter.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+ * Formats data for common licence data `/licences/{id}` page's
+ *
+ * @returns {Object} The data formatted for the view template
+ */
+function go (returnsData) {
+  const returns = _formatReturnToTableRow(returnsData.returns)
+
+  return {
+    activeTab: 'returns',
+    returnsUrl: 'return/internal',
```
Comment (robertparkinson) [inline app/presenters/licences/view-licence-returns.presenter.js:20]: is this ever going to be anything else? even so why not just hard code it in the nunjucks file until we need to make it dynamic?

**Review by robertparkinson — CHANGES_REQUESTED** [review]

File: test/controllers/licences.controller.test.js [hunk 10]
```diff
@@ -188,4 +189,42 @@ describe('Licences controller', () => {
...
+        expect(response.payload).to.contain('Status')
+      })
+    })
+
+    function _viewLicenceReturns () {
+      return {
+        activeTab: 'returns',
+        returns: [{}]
+      }
+    }
```
Comment (Cruikshanks) [inline test/controllers/licences.controller.test.js:228]: Not saying it's right, but currently, for any helper functions like this we put them outside the `describe()` blocks at the bottom of the file.

File: test/presenters/licences/view-licence-returns-presenter.test.js [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+const { describe, it, beforeEach } = exports.lab = Lab.script()
+const { expect } = Code
+
+// Thing under test
+const ViewLicenceReturnsPresenter = require('../../../app/presenters/licences/view-licence-returns.presenter')
+
+describe('View Licence returns presenter', () => {
+  let returnData
+  beforeEach(() => {
```
Comment (Cruikshanks) [inline test/presenters/licences/view-licence-returns-presenter.test.js:15]: NIT - whitespace

```suggestion
  let returnData

  beforeEach(() => {
```

File: test/services/licences/fetch-licence-returns.service.test.js [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+const LicenceHelper = require('../../support/helpers/licence.helper')
+
+describe('Fetch licence returns service', () => {
+  const licenceId = 'fef693fd-eb6f-478d-9f79-ab24749c5dc6'
+
+  beforeEach(async () => {
+    await DatabaseSupport.clean()
+  })
+
+  describe('when there is no optional data in the model', () => {
```
Comment (Cruikshanks) [inline test/services/licences/fetch-licence-returns.service.test.js:25]: Guessing this is a copy & paste description.

```suggestion
  describe('when the licence has return logs', () => {
```

File: test/services/licences/fetch-licence-returns.service.test.js [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    let firstReturn
+    let latestReturn
```
Comment (Cruikshanks) [inline test/services/licences/fetch-licence-returns.service.test.js:27]: I could be wrong but it looks like we are saving to these so we can access the IDs for asserting in the tests.

We have 2 alternate strategies for this. The first and easiest is just to specify the ID when you create the record

```javascript
await ReturnLogHelper.add({ ...returnData, id: '72c01307-1a63-48db-b59d-021e7bf623c6' })
```

The second is to ignore them on the basis we clear the table before inserting, so we know there are only these records in the table.

```javascript
  expect(result).to.equal({}, { skip: ['id] }
```

So, I think you can for go them. Else, we like to specify any `let`s _after_ the `const`s with a good ol' newline in between!

File: test/services/licences/view-licence-returns.service.test.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+const { describe, it, beforeEach, afterEach } = exports.lab = Lab.script()
+const { expect } = Code
+
+// Things we need to stub
+const ViewLicenceService = require('../../../app/services/licences/view-licence.service')
+const PaginatorPresenter = require('../../../app/presenters/paginator.presenter')
+const ViewLicenceReturnsPresenter = require('../../../app/presenters/licences/view-licence-returns.presenter')
+const FetchLicenceReturnsService = require('../../../app/services/licences/fetch-licence-returns.service')
+// Thing under test
```
Comment (Cruikshanks) [inline test/services/licences/view-licence-returns.service.test.js:16]: NIT - whitespace

```suggestion
const FetchLicenceReturnsService = require('../../../app/services/licences/fetch-licence-returns.service')

// Thing under test
```

File: test/services/licences/view-licence-returns.service.test.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+const ViewLicenceReturnsService = require('../../../app/services/licences/view-licence-returns.service')
+
+describe('View Licence service returns', () => {
+  const testId = '2c80bd22-a005-4cf4-a2a2-73812a9861de'
+  const page = 1
+  const auth = {}
+  const pagination = { page }
+  beforeEach(() => {
```
Comment (Cruikshanks) [inline test/services/licences/view-licence-returns.service.test.js:24]: NIT - whitespace

```suggestion
  const pagination = { page }

  beforeEach(() => {
```

File: test/services/licences/view-licence-returns.service.test.js [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+function _returnsFetch () {
+  return {
+    pagination: { total: 1 },
+    returns: []
+  }
+}
+
+function _licence () {
+  return { licenceName: 'fake licence' }
+}
```
Comment (Cruikshanks) [inline test/services/licences/view-licence-returns.service.test.js:69]: NIT - ordering

```suggestion
function _licence () {
  return { licenceName: 'fake licence' }
}

function _returnsFetch () {
  return {
    pagination: { total: 1 },
    returns: []
  }
}

function _returnsPresenter () {
  return {
    returnsUrl: 'return/internal',
    returns: [],
    activeTab: 'returns'
  }
}
```

**Review by Cruikshanks — CHANGES_REQUESTED** [review]
They are nearly all just nit-picks which hopefully soon ESLint will save me from highlighting (or more often missing!)

I've also seen @robertparkinson got in there before me and I concur with what he is saying.

File: app/presenters/licences/view-licence-returns.presenter.js [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+function _formatStatus (status) {
+  if (status === 'completed') return 'COMPLETE'
+  if (status === 'due') return 'OVERDUE'
+  return 'NO STATUS'
+}
+
+function _formatReturnToTableRow (returns) {
+  return returns.map(r => {
```
Comment (Cruikshanks) [inline app/presenters/licences/view-licence-returns.presenter.js:37]: SonarCloud and us - we always use brackets even when there is only 1 arg

```suggestion
  return returns.map((r) => {
```

**Review by Cruikshanks — CHANGES_REQUESTED** [review]
Some further tweaks. Went to check on SonarCloud. Happy to silence the method length stuff but it caught some things I should have spotted on the first pass.

File: app/presenters/licences/view-licence-returns.presenter.js (hunk 9 above)
Comment (jonathangoulding) [reply app/presenters/licences/view-licence-returns.presenter.js:20]: I will hard code in the njk

Comment (jonathangoulding) [reply app/presenters/licences/view-licence-returns.presenter.js]: done

File: test/controllers/licences.controller.test.js (hunk 10 above)
Comment (jonathangoulding) [reply test/controllers/licences.controller.test.js]: Done. Must have refactored at some point and missed this.

File: app/presenters/licences/view-licence-returns.presenter.js [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+
+function _formatPurpose (purpose) {
+  const [firstPurpose] = purpose
+  return firstPurpose.alias ? firstPurpose.alias : firstPurpose.tertiary.description
+}
+
+function _formatStatus (status) {
+  if (status === 'completed') return 'COMPLETE'
+  if (status === 'due') return 'OVERDUE'
+  return 'NO STATUS'
```
[the remark this answers is not in the document]
Comment (jonathangoulding) [reply app/presenters/licences/view-licence-returns.presenter.js]: done locally.

File: test/services/licences/fetch-licence-returns.service.test.js (hunk 13 above)
Comment (jonathangoulding) [reply test/services/licences/fetch-licence-returns.service.test.js]: Code updated.

This is needed:

```
      firstReturn.licenceRef = license.licenceRef
      latestReturn.licenceRef = license.licenceRef
```

As the licenceRef can't be overidden and forces `generateLicenceRef()`

File: test/services/licences/fetch-licence-returns.service.test.js [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+const { describe, it, beforeEach } = exports.lab = Lab.script()
+const { expect } = Code
+
+// Test helpers
+const DatabaseSupport = require('../../support/database.js')
+const FetchLicenceReturnsService = require('../../../app/services/licences/fetch-licence-returns.service')
+const ReturnLogHelper = require('../../support/helpers/return-log.helper.js')
+
+// Thing under test
+const LicenceHelper = require('../../support/helpers/licence.helper')
```
Comment (Cruikshanks) [inline test/services/licences/fetch-licence-returns.service.test.js:16]: Not quite!

```suggestion
// Test helpers
const DatabaseSupport = require('../../support/database.js')
const LicenceHelper = require('../../support/helpers/licence.helper')
const ReturnLogHelper = require('../../support/helpers/return-log.helper.js')

// Thing under test
const FetchLicenceReturnsService = require('../../../app/services/licences/fetch-licence-returns.service')
```

**Review by Cruikshanks — CHANGES_REQUESTED** [review]

**Review by Cruikshanks — APPROVED** [review]

**Pull request merged**